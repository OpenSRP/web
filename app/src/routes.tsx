import React from 'react';
import { DashboardOutlined, IdcardOutlined } from '@ant-design/icons';
import { INVENTORY_SERVICE_POINT_LIST_VIEW } from '@opensrp/inventory';
import {
  ACTIVE_PLANS_LIST_VIEW_URL,
  DRAFT_PLANS_LIST_VIEW_URL,
  COMPLETE_PLANS_LIST_VIEW_URL,
  TRASH_PLANS_LIST_VIEW_URL,
} from '@opensrp/plans';
import { isAuthorized } from '@opensrp/react-utils';
import { CATALOGUE_LIST_VIEW_URL } from '@opensrp/product-catalogue';
import ArchiveOutline from '@opensrp/ant-icons/lib/ArchiveOutline';
import MapMarkerOutline from '@opensrp/ant-icons/lib/MapMarkerOutline';
import {
  ENABLE_PLANS,
  ENABLE_INVENTORY,
  ENABLE_LOCATIONS,
  ENABLE_PRODUCT_CATALOGUE,
  ENABLE_TEAMS,
  ENABLE_FORM_CONFIGURATION,
  ENABLE_CARD_SUPPORT,
  OPENSRP_ROLES,
} from './configs/env';
import {
  ACTIVE,
  DRAFT,
  COMPLETE,
  TRASH,
  SERVICE_POINT_INVENTORY,
  USERS,
  USER_MANAGEMENT,
  LOCATION_UNIT,
  LOCATION_UNIT_GROUP,
  PRODUCT_CATALOGUE,
  TEAMS,
  FORM_CONFIGURATION,
  MANIFEST_RELEASES,
  DRAFT_FILES,
  TEAM_ASSIGNMENT,
  JSON_VALIDATORS,
  MISSIONS,
  ADMIN,
  INVENTORY,
} from './lang';

import {
  URL_USER,
  URL_LOCATION_UNIT,
  URL_LOCATION_UNIT_GROUP,
  URL_ADMIN,
  URL_TEAMS,
  URL_MANIFEST_RELEASE_LIST,
  URL_DRAFT_FILE_LIST,
  URL_TEAM_ASSIGNMENT,
  URL_JSON_VALIDATOR_LIST,
  URL_DOWNLOAD_CLIENT_DATA,
} from './constants';

/** Interface for menu items */
export interface Route {
  key: string;
  enabled?: boolean;
  url?: string;
  title: string;
  otherProps?: {
    icon?: string | JSX.Element;
  };
  children?: Route[];
}

// menu items schema
export const getRoutes = (roles: string[]): Route[] => {
  const activeRoles = OPENSRP_ROLES;
  return [
    {
      otherProps: { icon: <MapMarkerOutline className="sidebar-icons" /> },
      title: `${MISSIONS}`,
      key: 'missions',
      enabled:
        ENABLE_PLANS &&
        roles &&
        activeRoles.PLANS &&
        isAuthorized(roles, activeRoles.PLANS.split(',')),
      children: [
        { title: `${ACTIVE}`, url: `${ACTIVE_PLANS_LIST_VIEW_URL}`, key: 'missions-active' },
        { title: `${DRAFT}`, url: `${DRAFT_PLANS_LIST_VIEW_URL}`, key: 'missions-draft' },
        { title: `${COMPLETE}`, url: `${COMPLETE_PLANS_LIST_VIEW_URL}`, key: 'missions-complete' },
        { title: `${TRASH}`, url: `${TRASH_PLANS_LIST_VIEW_URL}`, key: 'missions-trash' },
      ],
    },
    {
      otherProps: { icon: <ArchiveOutline className="sidebar-icons" /> },
      title: `${INVENTORY}`,
      key: 'inventory',
      enabled:
        ENABLE_INVENTORY &&
        roles &&
        activeRoles.INVENTORY &&
        isAuthorized(roles, activeRoles.INVENTORY.split(',')),
      children: [
        {
          title: `${SERVICE_POINT_INVENTORY}`,
          url: `${INVENTORY_SERVICE_POINT_LIST_VIEW}`,
          key: 'inventory-list',
        },
      ],
    },
    {
      otherProps: { icon: <DashboardOutlined /> },
      title: `${ADMIN}`,
      key: 'admin',
      enabled: true,
      url: '/admin',
      children: [
        {
          title: `${USERS}`,
          key: 'users',
          enabled: roles && activeRoles.USERS && isAuthorized(roles, activeRoles.USERS.split(',')),
          children: [{ title: `${USER_MANAGEMENT}`, key: 'user-management', url: `${URL_USER}` }],
        },
        {
          title: 'Locations',
          key: 'location',
          enabled:
            ENABLE_LOCATIONS &&
            roles &&
            activeRoles.LOCATIONS &&
            isAuthorized(roles, activeRoles.LOCATIONS.split(',')),
          children: [
            { title: `${LOCATION_UNIT}`, url: `${URL_LOCATION_UNIT}`, key: 'location-unit' },
            {
              title: `${LOCATION_UNIT_GROUP}`,
              url: `${URL_LOCATION_UNIT_GROUP}`,
              key: 'location-group',
            },
          ],
        },
        {
          title: `${PRODUCT_CATALOGUE}`,
          key: 'product-catalogue',
          enabled:
            ENABLE_PRODUCT_CATALOGUE &&
            roles &&
            activeRoles.PRODUCT_CATALOGUE &&
            isAuthorized(roles, activeRoles.PRODUCT_CATALOGUE.split(',')),
          url: `${URL_ADMIN}${CATALOGUE_LIST_VIEW_URL}`,
        },
        {
          title: `${TEAMS}`,
          key: 'teams',
          enabled:
            ENABLE_TEAMS &&
            roles &&
            activeRoles.TEAMS &&
            isAuthorized(roles, activeRoles.TEAMS.split(',')),
          children: [
            { title: `${TEAMS}`, url: `${URL_TEAMS}`, key: 'teams-list' },
            {
              title: `${TEAM_ASSIGNMENT}`,
              url: `${URL_TEAM_ASSIGNMENT}`,
              key: 'team-assignment',
            },
          ],
        },
        {
          title: `${FORM_CONFIGURATION}`,
          key: 'form-config',
          enabled:
            ENABLE_FORM_CONFIGURATION &&
            roles &&
            activeRoles.FORM_CONFIGURATION &&
            isAuthorized(roles, activeRoles.FORM_CONFIGURATION.split(',')),
          children: [
            {
              title: `${MANIFEST_RELEASES}`,
              key: 'form-config-releases',
              url: `${URL_MANIFEST_RELEASE_LIST}`,
            },
            { title: `${DRAFT_FILES}`, key: 'form-config-draft', url: `${URL_DRAFT_FILE_LIST}` },
            {
              title: `${JSON_VALIDATORS}`,
              key: 'form-config-validators',
              url: `${URL_JSON_VALIDATOR_LIST}`,
            },
          ],
        },
      ],
    },
    {
      otherProps: { icon: <IdcardOutlined /> },
      title: 'Card Support',
      key: 'card-support',
      enabled:
        ENABLE_CARD_SUPPORT &&
        roles &&
        activeRoles.CARD_SUPPORT &&
        isAuthorized(roles, activeRoles.CARD_SUPPORT.split(',')),
      children: [
        {
          title: 'Download Client Data',
          url: `${URL_DOWNLOAD_CLIENT_DATA}`,
          key: 'download-client-data',
        },
      ],
    },
  ];
};
