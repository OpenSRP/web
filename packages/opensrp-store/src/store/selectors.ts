import { reducerName as sessionReducerName } from '@onaio/session-reducer';
import { Store } from 'redux';
import { Dictionary } from '@onaio/utils';
import { createSelector } from 'reselect';

/** get API Token from the Redux store
 *
 * @param {any} state - the redux store
 */
interface StoreState {
  session: {
    [key: string]: Dictionary;
  };
}

interface APIFilters {
  accessToken?: boolean;
  apiToken?: boolean;
  providerState?: boolean;
}

/** Fetch access token from store if access_token filter val is true
 *
 * @param _ - redux store
 * @param props - api filters object
 * @returns {boolean}
 */

export const fetchAccessToken = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.accessToken || null;

/** Fetch api token from store if api_token filter val is true
 *
 * @param _ redux store
 * @param props api filters object
 * @returns {boolean}
 */

export const fetchApiToken = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.apiToken || null;

/** Fetch oauth provider state from store if providerState filter val is true
 *
 * @param _ redux store
 * @param props api filters object
 * @returns {boolean}
 */

export const fetchOauthProviderState = (_: Partial<Store>, props: APIFilters): boolean | null =>
  props.providerState || null;

/** Gets extra data object from store
 *
 * @param state - redux store
 */

export const getExtraData = (state: Partial<Store>): Dictionary =>
  (state as StoreState)[sessionReducerName].extraData;

/**
 * @param {any} state - the redux store
 * @returns {Object} - the states
 */

export const makeAPIStateSelector = () =>
  createSelector(
    [fetchAccessToken, fetchApiToken, fetchOauthProviderState, getExtraData],
    (accessToken, apiToken, oauthState, extraData) => {
      if (accessToken) {
        return (extraData.oAuth2Data && extraData.oAuth2Data.access_token) || null;
      } else if (apiToken) {
        return extraData.api_token || null;
      } else if (oauthState) {
        return (extraData.oAuth2Data && extraData.oAuth2Data.state) || null;
      }
    }
  );
