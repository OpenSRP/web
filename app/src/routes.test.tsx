import { store } from '@opensrp/store';
import { authenticateUser } from '@onaio/session-reducer';
import { getRoutes } from './routes';
jest.mock('./configs/env');

describe('routes', () => {
  it('Test routes only return enabled values', () => {
    const routes = getRoutes([]);

    expect(routes).toMatchInlineSnapshot(`
      Array [
        Object {
          "children": Array [],
          "enabled": true,
          "key": "admin",
          "otherProps": Object {
            "icon": <ForwardRef(DashboardOutlined) />,
          },
          "title": "Admin",
          "url": "/admin",
        },
      ]
    `);
  });

  it('Test Enabled Routes', () => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        {
          roles: ['ROLE_VIEW_KEYCLOAK_USERS'],
          username: 'superset-user',
          user_id: 'cab07278-c77b-4bc7-b154-bcbf01b7d35b',
        }
      )
    );

    const envModule = require('./configs/env');
    envModule.ENABLE_LOCATIONS = 'true';
    envModule.ENABLE_TEAMS = 'true';
    envModule.ENABLE_INVENTORY = 'true';
    envModule.ENABLE_FORM_CONFIGURATION = 'true';
    envModule.ENABLE_PRODUCT_CATALOGUE = 'true';
    envModule.ENABLE_PLANS = 'true';
    envModule.ENABLE_CARD_SUPPORT = 'true';
    envModule.OPENSRP_ROLES = {
      USERS: 'ROLE_EDIT_KEYCLOAK_USERS',
      PLANS: 'ROLE_VIEW_KEYCLOAK_USERS',
      LOCATIONS: 'ROLE_VIEW_KEYCLOAK_USERS',
      CARD_SUPPORT: 'ROLE_VIEW_KEYCLOAK_USERS',
      INVENTORY: 'ROLE_VIEW_KEYCLOAK_USERS',
      TEAMS: 'ROLE_VIEW_KEYCLOAK_USERS',
      PRODUCT_CATALOGUE: 'ROLE_VIEW_KEYCLOAK_USERS',
      FORM_CONFIGURATION: 'ROLE_VIEW_KEYCLOAK_USERS',
    };
    const routes = getRoutes(['ROLE_EDIT_KEYCLOAK_USERS', 'ROLE_VIEW_KEYCLOAK_USERS']);

    expect(routes).toMatchSnapshot();
  });
});
