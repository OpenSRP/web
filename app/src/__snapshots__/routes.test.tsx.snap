// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes Test Enabled Routes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "key": "missions-active",
        "title": "Active",
        "url": "/missions/active",
      },
      Object {
        "key": "missions-draft",
        "title": "Draft",
        "url": "/missions/draft",
      },
      Object {
        "key": "missions-complete",
        "title": "Complete",
        "url": "/missions/complete",
      },
      Object {
        "key": "missions-trash",
        "title": "Retired",
        "url": "/missions/trash",
      },
    ],
    "enabled": true,
    "key": "missions",
    "otherProps": Object {
      "icon": <MapMarkerOutlineIcon
        className="sidebar-icons"
      />,
    },
    "title": "Missions",
  },
  Object {
    "children": Array [
      Object {
        "key": "inventory-list",
        "title": "Service point inventory",
        "url": "/inventory/list",
      },
    ],
    "enabled": true,
    "key": "inventory",
    "otherProps": Object {
      "icon": <ArchiveOutlineIcon
        className="sidebar-icons"
      />,
    },
    "title": "Inventory",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "key": "user-management",
            "title": "User Management",
            "url": "/admin/users",
          },
        ],
        "enabled": true,
        "key": "users",
        "title": "Users",
      },
      Object {
        "children": Array [
          Object {
            "key": "location-unit",
            "title": "Location unit",
            "url": "/admin/location/unit",
          },
          Object {
            "key": "location-group",
            "title": "Location unit group",
            "url": "/admin/location/group",
          },
        ],
        "enabled": true,
        "key": "location",
        "title": "Locations",
      },
      Object {
        "enabled": true,
        "key": "product-catalogue",
        "title": "Product Catalogue",
        "url": "/admin/admin/product-catalogue",
      },
      Object {
        "children": Array [
          Object {
            "key": "teams-list",
            "title": "Teams",
            "url": "/admin/teams",
          },
          Object {
            "key": "team-assignment",
            "title": "Team Assignment",
            "url": "/admin/teams/team-assignment",
          },
        ],
        "enabled": true,
        "key": "teams",
        "title": "Teams",
      },
      Object {
        "children": Array [
          Object {
            "key": "form-config-releases",
            "title": "Manifest Releases",
            "url": "/admin/form-config/releases",
          },
          Object {
            "key": "form-config-draft",
            "title": "Draft Files",
            "url": "/admin/form-config/drafts",
          },
          Object {
            "key": "form-config-validators",
            "title": "JSON Validators",
            "url": "/admin/form-config/json-validators",
          },
        ],
        "enabled": true,
        "key": "form-config",
        "title": "Form Configuration",
      },
    ],
    "enabled": true,
    "key": "admin",
    "otherProps": Object {
      "icon": <ForwardRef(DashboardOutlined) />,
    },
    "title": "Admin",
    "url": "/admin",
  },
  Object {
    "children": Array [
      Object {
        "key": "download-client-data",
        "title": "Download Client Data",
        "url": "/card-support/download-client-data",
      },
    ],
    "enabled": true,
    "key": "card-support",
    "otherProps": Object {
      "icon": <ForwardRef(IdcardOutlined) />,
    },
    "title": "Card Support",
  },
]
`;
